### DISCUSSION TITLE
Getting Source Timecode of Asset Clip in Timeline

### DISCUSSION THREAD

**Discussion by latenitefilms** [orchetect/swift-daw-file-tools#28] [Q&A]
Following on from the discussion in https://github.com/TheAcharya/MarkersExtractor/issues/81, I'd love to work out how to test `DAWFileKit` in a very specific situation so that I can update the FCP Cafe explanation/documentation [here](https://fcp.cafe/developers/fcpxml/#getting-source-timecode).

If I have the following FCPXML:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE fcpxml>

<fcpxml version="1.11">
    <resources>
        <format id="r1" name="FFVideoFormat1080p24" frameDuration="100/2400s" width="1920" height="1080" colorSpace="1-1-1 (Rec. 709)"/>
        <asset id="r2" name="Test Clip - 25fps - 01-02-03-04" uid="F03C4D168FE7934996E6408D973FFB97" start="9307900/2500s" duration="60s" hasVideo="1" format="r3" hasAudio="1" videoSources="1" audioSources="1" audioChannels="2" audioRate="48000">
            <media-rep kind="original-media" sig="F03C4D168FE7934996E6408D973FFB97" src="file:///Folder/Test%20Clip%20-%2025fps%20-%2001-02-03-04.mov"/>
        </asset>
        <format id="r3" name="FFVideoFormat1080p25" frameDuration="100/2500s" width="1920" height="1080" colorSpace="1-1-1 (Rec. 709)"/>
    </resources>
    <project name="24fps Timeline" uid="99AB3F85-FDA0-49CD-83BB-B663A0F6FEA9" modDate="2024-02-10 06:02:11 +1100">
        <sequence format="r1" duration="12000/1920s" tcStart="7148480/1920s" tcFormat="NDF" audioLayout="stereo" audioRate="48k">
            <spine>
                <asset-clip ref="r2" offset="7148480/1920s" name="Test Clip - 25fps - 01-02-03-04" start="93079/25s" duration="12000/1920s" format="r3" tcFormat="NDF" audioRole="dialogue">
                    <conform-rate srcFrameRate="25"/>
                </asset-clip>
            </spine>
        </sequence>
    </project>
</fcpxml>
```

How can I use `DAWFileKit` to get the source timecode of the `asset-clip` at the timeline/project time of: 

> `01:02:08:07` / 3728.29 seconds / 89479 frames

Looking in Final Cut Pro's Source Timecode display (and my burnt-in timecode) I know the answer is: 

> `01:02:08:02` / 3728.08 seconds / 93202 frames

How can I reproduce this using DAWFileKit?

**Comment by orchetect** [comment]
I don't quite understand the question. Do you mean you want to know what clip intersects an arbitrary timecode?

**Reply by latenitefilms** [reply]
Sorry, just saw this. Yes, I want to feed in an arbitrary timeline/project timecode value, and get the clip's timecode value at that frame.

**Reply by orchetect** [reply]
Not sure I follow yet. Given an absolute timecode, you want to know what the `asset-clip`'s local timeline timecode is for it?

Since this example has different frame rates (24fps project, 25fps asset), you can read timecode from DAWFileKit either using the top-level frame rate (project rate) or the clip's local frame rate.

DAWFileKit's default is to use the top-level timeline's context. So it would give you timecode as 24fps by default.

**Reply by latenitefilms** [reply]
Basically when I originally posted this, and was adding the documentation to [FCP Cafe](https://fcp.cafe/developers/fcpxml/#getting-source-timecode), I was attempting to use `DAWFileKit` for my [Metaburner](https://metaburner.pro) project. In the end we decided not to go down the `DAWFileKit` route, and use our own proprietary `FCPXMLKit` implementation - which the jury is still out on whether or not that was a silly idea, as we've basically had to go through all the same chaos and problem-solving as you've had to solve in `DAWFileKit`.

What we do is essentially loop through every frame on the timeline/project, and collect information about all the clips at that frame.

In this case, I wanted to "feed in" the timeline/project value (i.e. `01:02:08:07` - although we'd most likely be feeding in `CMTime` of a fractional value, as opposed to "timecode"), and "collect" all the metadata for the clips at that timeline/project frame. In this case, we only have a single `asset-clip`, and we want to get the clip's source/local timecode value in it's source/local frame rate - i.e. `01:02:08:02 @ 25fps`.

The reason this example has a 24fps timeline/project and a 25fps clip is because I wanted to make sure we're actually getting the clip's local/source timecode at the clip's local/source frame-rate.

We've since got this working pretty well in our own `FCPXMLKit` - however there's definitely still edge cases where things are a bit funky, for example, in the below screenshot the timecode is perfect, EXECPT for this "segment" that has **Normal 100%** speed, the source timecode is one frame off.

![image](https://github.com/orchetect/DAWFileKit/assets/22286696/864c10e4-8543-4f01-ac15-cff5e3a00762)

**Reply by orchetect** [reply]
Certainly possible with DAWFileKit but I'd probably have to built out a new extraction paradigm based on time and not objects. Basically something that could build an internal model of time ranges and objects in those ranges with a given timeline context (which would be main timeline by default). Then it would be trivial to feed it an arbitrary time and get back everything that intersects at that time. I could see it being a useful feature.

**Reply by latenitefilms** [reply]
That would be amazing!

I believe that's kinda what we're doing now - we have `FCPXMLReport`, we adds reporting capabilities to our `FCPXMLKit` framework.

For me to consider switching to `DAWFileKit` though would require a lot more additions to `DAWFileKit` because we need to access video inspector parameters (taking into account keyframes), stabilisation, etc. `DAWFileKit` would basically need to have access to everything that the current DTD supports, which is a lot of work.

I'm also not sure `DAWFileKit` takes into account speed changes (`timeMap`) in its current implementation, which is required (and bloody tricky!) when trying to get source timecode from a Multicam Clip with a speed effect, which contains a Sync Clip with a speed effect which contains an Asset Clip with a speed effect, etc (let alone the `conform-rate` stuff too!).

FCPXML is the devil, haha.

**Reply by orchetect** [reply]
That's totally fine. Like a lot of open-source, DAWFileKit is a labor of love and new features are generally added as-needed. Use whatever works for you.

> not sure DAWFileKit takes into account speed changes (timeMap) in its current implementation

No, it's on the list of things to-do of course.

I like the notion of making clear distinction between the XML model and reasoning/reporting on the data. Right now all the API intermingle in the library but I can see making that more clear.
